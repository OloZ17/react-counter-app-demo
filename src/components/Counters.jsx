import React from 'react';
import Counter from './Counter';
import { func, array } from 'prop-types';

const Counters = (props) => {
  const { counters, onReset, onIncrement, onDelete, onDecrement, onRestart } =
    props;
  return (
    <div className="container">
      <div className="row justify-content-center">
        <div className="col-4 text-center">
          <button
            data-testid="reset-button"
            className="btn btn-success m-2"
            onClick={onReset}
            disabled={counters.length === 0 ? 'disabled' : ''}
          >
            <i className="fa fa-refresh" aria-hidden="true" />
          </button>
          <button
            data-testid="restart-button"
            className="btn btn-primary m-2"
            onClick={onRestart}
            disabled={counters.length !== 0 ? 'disabled' : ''}
          >
            <i className="fa fa-recycle" aria-hidden="true" />
          </button>
        </div>
      </div>
      {counters.map((counter) => (
        <Counter
          key={counter.id}
          counter={counter}
          onIncrement={onIncrement}
          onDecrement={onDecrement}
          onDelete={onDelete}
        />
      ))}
    </div>
  );
};

Counters.propTypes = {
  onIncrement: func,
  onDecrement: func,
  onReset: func,
  onDelete: func,
  onRestart: func,
  counters: array,
};

export default Counters;
