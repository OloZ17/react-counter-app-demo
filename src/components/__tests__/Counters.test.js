import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Counters from '../Counters';

describe('Testing Counters() Component', () => {
  const counters = [
    { id: 1, value: 0 },
    { id: 2, value: 0 },
    { id: 3, value: 0 },
    { id: 4, value: 0 },
  ];

  const onReset = jest.fn();
  const onRestart = jest.fn();

  it('should take a snapshot', () => {
    const { asFragment } = render(
      <Counters onReset={onReset} onRestart={onRestart} counters={counters} />
    );
    expect(asFragment(<Counters />)).toMatchSnapshot();
  });

  it('should render buttons with right attributes with counter.length > 0 and should click the right button', () => {
    render(
      <Counters onReset={onReset} onRestart={onRestart} counters={counters} />
    );
    expect(screen.getByTestId('reset-button')).not.toHaveAttribute('disabled');
    expect(screen.getByTestId('restart-button')).toHaveAttribute('disabled');
    const toOnReset = screen.getByTestId('reset-button');
    fireEvent.click(toOnReset);
    expect(onReset).toBeCalled();
  });

  it('should render buttons with right attributes with counter.length < 1 and should click the right button', () => {
    const counters = [];
    render(
      <Counters onReset={onReset} onRestart={onRestart} counters={counters} />
    );
    expect(screen.getByTestId('reset-button')).toHaveAttribute('disabled');
    expect(screen.getByTestId('restart-button')).not.toHaveAttribute(
      'disabled'
    );
    const toOnRestart = screen.getByTestId('restart-button');
    fireEvent.click(toOnRestart);
    expect(onRestart).toBeCalled();
  });
});
