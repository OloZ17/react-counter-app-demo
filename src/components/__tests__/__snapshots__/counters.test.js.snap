// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Counters() Component should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="row justify-content-center"
    >
      <div
        class="col-4 text-center"
      >
        <button
          class="btn btn-success m-2"
          data-testid="reset-button"
        >
          <i
            aria-hidden="true"
            class="fa fa-refresh"
          />
        </button>
        <button
          class="btn btn-primary m-2"
          data-testid="restart-button"
          disabled=""
        >
          <i
            aria-hidden="true"
            class="fa fa-recycle"
          />
        </button>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-4 text-center"
        >
          <span
            class="badge m-2 bg-warning"
            data-testid="total-counter"
            style="font-size: 21px; padding: 9px; vertical-align: middle;"
          >
            Zero
          </span>
          <button
            class="btn btn-secondary"
            data-testid="increment-button"
          >
            <i
              aria-hidden="true"
              class="fa fa-plus-circle"
            />
          </button>
          <button
            class="btn btn-info m-2"
            data-testid="decrement-button"
            disabled=""
          >
            <i
              aria-hidden="true"
              class="fa fa-minus-circle"
            />
          </button>
          <button
            class="btn btn-danger"
            data-testid="delete-button"
          >
            <i
              aria-hidden="true"
              class="fa fa-trash-o"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-4 text-center"
        >
          <span
            class="badge m-2 bg-warning"
            data-testid="total-counter"
            style="font-size: 21px; padding: 9px; vertical-align: middle;"
          >
            Zero
          </span>
          <button
            class="btn btn-secondary"
            data-testid="increment-button"
          >
            <i
              aria-hidden="true"
              class="fa fa-plus-circle"
            />
          </button>
          <button
            class="btn btn-info m-2"
            data-testid="decrement-button"
            disabled=""
          >
            <i
              aria-hidden="true"
              class="fa fa-minus-circle"
            />
          </button>
          <button
            class="btn btn-danger"
            data-testid="delete-button"
          >
            <i
              aria-hidden="true"
              class="fa fa-trash-o"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-4 text-center"
        >
          <span
            class="badge m-2 bg-warning"
            data-testid="total-counter"
            style="font-size: 21px; padding: 9px; vertical-align: middle;"
          >
            Zero
          </span>
          <button
            class="btn btn-secondary"
            data-testid="increment-button"
          >
            <i
              aria-hidden="true"
              class="fa fa-plus-circle"
            />
          </button>
          <button
            class="btn btn-info m-2"
            data-testid="decrement-button"
            disabled=""
          >
            <i
              aria-hidden="true"
              class="fa fa-minus-circle"
            />
          </button>
          <button
            class="btn btn-danger"
            data-testid="delete-button"
          >
            <i
              aria-hidden="true"
              class="fa fa-trash-o"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="container"
    >
      <div
        class="row justify-content-center"
      >
        <div
          class="col-4 text-center"
        >
          <span
            class="badge m-2 bg-warning"
            data-testid="total-counter"
            style="font-size: 21px; padding: 9px; vertical-align: middle;"
          >
            Zero
          </span>
          <button
            class="btn btn-secondary"
            data-testid="increment-button"
          >
            <i
              aria-hidden="true"
              class="fa fa-plus-circle"
            />
          </button>
          <button
            class="btn btn-info m-2"
            data-testid="decrement-button"
            disabled=""
          >
            <i
              aria-hidden="true"
              class="fa fa-minus-circle"
            />
          </button>
          <button
            class="btn btn-danger"
            data-testid="delete-button"
          >
            <i
              aria-hidden="true"
              class="fa fa-trash-o"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
